lib/cequel/metal/statement.rb

```diff
@@ -10,16 +10,30 @@ module Metal
     class Statement
       # @return [Array] bind variables for CQL string
       attr_reader :bind_vars
+      # @return [Array] cassandra type hints for bind variables
 
-      def initialize
-        @cql, @bind_vars = [], []
+      def initialize(cql_or_prepared='', bind_vars=[])
+        cql, prepared = *case cql_or_prepared
+                         when Cassandra::Statements::Prepared
+                           [cql_or_prepared.cql, cql_or_prepared]
+                         else
+                           [cql_or_prepared.to_s, nil]
+                         end
+
+        @cql, @prepared, @bind_vars = cql, prepared, bind_vars
       end
 
       #
       # @return [String] CQL statement
       #
-      def cql
-        @cql.join
+      def to_s
+        @cql
+      end
+      alias_method :cql, :to_s
+
+      # @return [Cassandra::Statements::Prepared] prepared version of this statement
+      def prepare(keyspace)
+        @prepared ||= keyspace.client.prepare(cql)
```

Comment:
I think my previous comment on this was canceled somehow but...
Memoizing the prepared CQL is good here but each `Cequel::Metal::Statement` object will ask the server for  a new prepared request even if this statement was already prepared once.

To mitigate this I cherry-picked my `PreparedStatementManager` which reuses prepared statements per `Cequel::Metal::Statement` instance. Right now the `Keyspace` only has the setter/getter for it. We'll need to use `Keyspace#prepare_statement` this during execute instead and remove `Cequel::Metal::Statement#prepare`